In [117]:
import pandas as pd
import numpy as np
data = pd.read_csv("audit_risk.csv")
data.head(3)

Sector_score LOCATION_ID  PARA_A  Score_A  Risk_A  PARA_B  Score_B  Risk_B  \
0          3.89          23    4.18      0.6   2.508    2.50      0.2   0.500   
1          3.89           6    0.00      0.2   0.000    4.83      0.2   0.966   
2          3.89           6    0.51      0.2   0.102    0.23      0.2   0.046   

   TOTAL  numbers  ...  RiSk_E  History  Prob  Risk_F  Score  Inherent_Risk  \
0   6.68      5.0  ...     0.4        0   0.2     0.0    2.4          8.574   
1   4.83      5.0  ...     0.4        0   0.2     0.0    2.0          2.554   
2   0.74      5.0  ...     0.4        0   0.2     0.0    2.0          1.548   

   CONTROL_RISK  Detection_Risk  Audit_Risk  Risk  
0           0.4             0.5      1.7148     1  
1           0.4             0.5      0.5108     0  
2           0.4             0.5      0.3096     0  

[3 rows x 27 columns]

In [118]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 776 entries, 0 to 775
Data columns (total 27 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Sector_score    776 non-null    float64
 1   LOCATION_ID     776 non-null    object 
 2   PARA_A          776 non-null    float64
 3   Score_A         776 non-null    float64
 4   Risk_A          776 non-null    float64
 5   PARA_B          776 non-null    float64
 6   Score_B         776 non-null    float64
 7   Risk_B          776 non-null    float64
 8   TOTAL           776 non-null    float64
 9   numbers         776 non-null    float64
 10  Score_B.1       776 non-null    float64
 11  Risk_C          776 non-null    float64
 12  Money_Value     775 non-null    float64
 13  Score_MV        776 non-null    float64
 14  Risk_D          776 non-null    float64
 15  District_Loss   776 non-null    int64  
 16  PROB            776 non-null    float64
 17  RiSk_E          776 non-null    flo

Всего есть 25 признаков и 1 целевая переменная (бинарная) - нужно определить является ли организация мошеннической или нет

In [119]:
print(data.shape)

(776, 27)


###### Всего 776 организаций

Посмотрим на соотношение классов

In [120]:
data.iloc[:, -1].value_counts()

0    471
1    305
Name: Risk, dtype: int64

Стандартизируем признаки

In [121]:
from sklearn.preprocessing import StandardScaler

In [122]:
data.columns.tolist()

['Sector_score',
 'LOCATION_ID',
 'PARA_A',
 'Score_A',
 'Risk_A',
 'PARA_B',
 'Score_B',
 'Risk_B',
 'TOTAL',
 'numbers',
 'Score_B.1',
 'Risk_C',
 'Money_Value',
 'Score_MV',
 'Risk_D',
 'District_Loss',
 'PROB',
 'RiSk_E',
 'History',
 'Prob',
 'Risk_F',
 'Score',
 'Inherent_Risk',
 'CONTROL_RISK',
 'Detection_Risk',
 'Audit_Risk',
 'Risk']

In [123]:
feature_names = ['Sector_score', 'LOCATION_ID', 'PARA_A', 'Score_A', 'Risk_A', 'PARA_B', 'Score_B', 'Risk_B', 'TOTAL', 'numbers',
                 'Score_B.1', 'Risk_C', 'Money_Value', 'Score_MV', 'Risk_D', 'District_Loss', 'PROB', 'RiSk_E', 'History', 'Prob',
                  'Risk_F', 'Score', 'Inherent_Risk', 'CONTROL_RISK', 'Detection_Risk', 'Audit_Risk']

target_name = 'Risk'

In [124]:
data = data[feature_names + [target_name]]
data.head(2)

Sector_score LOCATION_ID  PARA_A  Score_A  Risk_A  PARA_B  Score_B  Risk_B  \
0          3.89          23    4.18      0.6   2.508    2.50      0.2   0.500   
1          3.89           6    0.00      0.2   0.000    4.83      0.2   0.966   

   TOTAL  numbers  ...  RiSk_E  History  Prob  Risk_F  Score  Inherent_Risk  \
0   6.68      5.0  ...     0.4        0   0.2     0.0    2.4          8.574   
1   4.83      5.0  ...     0.4        0   0.2     0.0    2.0          2.554   

   CONTROL_RISK  Detection_Risk  Audit_Risk  Risk  
0           0.4             0.5      1.7148     1  
1           0.4             0.5      0.5108     0  

[2 rows x 27 columns]

In [125]:
feature_names_for_stand = data[feature_names].select_dtypes(include=['float64']).columns.tolist()

scaler = StandardScaler()
stand_features = scaler.fit_transform(data[feature_names_for_stand])

data[feature_names_for_stand] = pd.DataFrame(stand_features, columns=feature_names_for_stand)

data.head()

Sector_score LOCATION_ID    PARA_A   Score_A    Risk_A        PARA_B  \
0     -0.670465          23  0.304800  1.429846  0.336502 -1.658295e-01   
1     -0.670465           6 -0.431736 -0.869761 -0.392943 -1.192773e-01   
2     -0.670465           6 -0.341872 -0.869761 -0.363277 -2.111829e-01   
3     -0.670465           6 -0.431736 -0.869761 -0.392943  2.317208e-07   
4     -0.670465           6 -0.431736 -0.869761 -0.392943 -2.141798e-01   

    Score_B    Risk_B     TOTAL   numbers  ...    RiSk_E  History      Prob  \
0 -0.666752 -0.194121 -0.127506 -0.255998  ... -0.410417        0 -0.246568   
1 -0.666752 -0.178615 -0.163583 -0.255998  ... -0.410417        0 -0.246568   
2 -0.666752 -0.209227 -0.243341 -0.255998  ... -0.410417        0 -0.246568   
3  1.690422  0.004858 -0.047162  3.527894  ... -0.410417        0 -0.246568   
4 -0.666752 -0.210226 -0.256212 -0.255998  ... -0.410417        0 -0.246568   

     Risk_F     Score  Inherent_Risk  CONTROL_RISK  Detection_Risk  \
0 -0.175398 -0.352503      -0.166468     -0.388662             0.0   
1 -0.175398 -0.818503      -0.276513     -0.388662             0.0   
2 -0.175398 -0.818503      -0.294902     -0.388662             0.0   
3 -0.175398  1.977497      -0.002753     -0.388662             0.0   
4 -0.175398 -0.818503      -0.297315     -0.388662             0.0   

   Audit_Risk  Risk  
0   -0.141123     1  
1   -0.172280     0  
2   -0.177487     0  
3   -0.094770     1  
4   -0.178170     0  

[5 rows x 27 columns]

In [126]:
data = data.drop('LOCATION_ID', axis=1)

In [127]:
display(data.isna().sum())

Sector_score      0
PARA_A            0
Score_A           0
Risk_A            0
PARA_B            0
Score_B           0
Risk_B            0
TOTAL             0
numbers           0
Score_B.1         0
Risk_C            0
Money_Value       1
Score_MV          0
Risk_D            0
District_Loss     0
PROB              0
RiSk_E            0
History           0
Prob              0
Risk_F            0
Score             0
Inherent_Risk     0
CONTROL_RISK      0
Detection_Risk    0
Audit_Risk        0
Risk              0
dtype: int64

In [128]:
data['Money_Value'] = data['Money_Value'].replace(np.nan, 0)
display(data.isna().sum())

Sector_score      0
PARA_A            0
Score_A           0
Risk_A            0
PARA_B            0
Score_B           0
Risk_B            0
TOTAL             0
numbers           0
Score_B.1         0
Risk_C            0
Money_Value       0
Score_MV          0
Risk_D            0
District_Loss     0
PROB              0
RiSk_E            0
History           0
Prob              0
Risk_F            0
Score             0
Inherent_Risk     0
CONTROL_RISK      0
Detection_Risk    0
Audit_Risk        0
Risk              0
dtype: int64

Обучаем DecisionTreeClassifier

In [129]:
from sklearn.model_selection import train_test_split

x_data = data.iloc[:,:-1]
y_data = data.iloc[:,-1]

x_train, x_test, y_train, y_test = train_test_split(x_data, y_data, test_size=0.25, random_state=42)

In [130]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import recall_score

classifier = DecisionTreeClassifier()
classifier.fit(x_train, y_train)
y_predict = classifier.predict(x_test)

Проверяем качество

In [131]:
from sklearn.metrics import recall_score, precision_score, roc_auc_score, accuracy_score, f1_score

def evaluate_results(y_test, y_predict):
    print('Classification results:')
    f1 = f1_score(y_test, y_predict)
    print("f1: %.2f%%" % (f1 * 100.0)) 
    roc = roc_auc_score(y_test, y_predict)
    print("roc: %.2f%%" % (roc * 100.0)) 
    rec = recall_score(y_test, y_predict, average='binary')
    print("recall: %.2f%%" % (rec * 100.0)) 
    prc = precision_score(y_test, y_predict, average='binary')
    print("precision: %.2f%%" % (prc * 100.0)) 

    
evaluate_results(y_test, y_predict)

Classification results:
f1: 100.00%
roc: 100.00%
recall: 100.00%
precision: 100.00%


### Теперь очередь за PU learning

Представим, что нам неизвестны негативы и часть позитивов

In [132]:
mod_data = data.copy()
#get the indices of the positives samples
pos_ind = np.where(mod_data.iloc[:,-1].values == 1)[0]
#shuffle them
np.random.shuffle(pos_ind)
# leave just 25% of the positives marked
pos_sample_len = int(np.ceil(0.25 * len(pos_ind)))
print(f'Using {pos_sample_len}/{len(pos_ind)} as positives and unlabeling the rest')
pos_sample = pos_ind[:pos_sample_len]

Using 77/305 as positives and unlabeling the rest


Создаем столбец для новой целевой переменной, где у нас два класса - P (1) и U (-1)

In [133]:
mod_data['class_test'] = -1
mod_data.loc[pos_sample,'class_test'] = 1
print('target variable:\n', mod_data.iloc[:,-1].value_counts())

target variable:
 -1    699
 1     77
Name: class_test, dtype: int64


* We now have just 699 positive samples labeled as 1 in the 'class_test' col while the rest is unlabeled as -1.

In [134]:
mod_data.head(4)

Sector_score    PARA_A   Score_A    Risk_A        PARA_B   Score_B  \
0     -0.670465  0.304800  1.429846  0.336502 -1.658295e-01 -0.666752   
1     -0.670465 -0.431736 -0.869761 -0.392943 -1.192773e-01 -0.666752   
2     -0.670465 -0.341872 -0.869761 -0.363277 -2.111829e-01 -0.666752   
3     -0.670465 -0.431736 -0.869761 -0.392943  2.317208e-07  1.690422   

     Risk_B     TOTAL   numbers  Score_B.1  ...  History      Prob    Risk_F  \
0 -0.194121 -0.127506 -0.255998  -0.295285  ...        0 -0.246568 -0.175398   
1 -0.178615 -0.163583 -0.255998  -0.295285  ...        0 -0.246568 -0.175398   
2 -0.209227 -0.243341 -0.255998  -0.295285  ...        0 -0.246568 -0.175398   
3  0.004858 -0.047162  3.527894   4.686039  ...        0 -0.246568 -0.175398   

      Score  Inherent_Risk  CONTROL_RISK  Detection_Risk  Audit_Risk  Risk  \
0 -0.352503      -0.166468     -0.388662             0.0   -0.141123     1   
1 -0.818503      -0.276513     -0.388662             0.0   -0.172280     0   
2 -0.818503      -0.294902     -0.388662             0.0   -0.177487     0   
3  1.977497      -0.002753     -0.388662             0.0   -0.094770     1   

   class_test  
0          -1  
1          -1  
2          -1  
3          -1  

[4 rows x 27 columns]

Remember that this data frame (x_data) includes the former target variable that we keep here just to compare the results

[:-2] is the original class label for positive and negative data [:-1] is the new class for positive and unlabeled data

In [135]:
x_data = mod_data.iloc[:,:-2].values # just the X 
y_labeled = mod_data.iloc[:,-1].values # new class (just the P & U)
y_positive = mod_data.iloc[:,-2].values # original class

### 1. random negative sampling

In [136]:
mod_data = mod_data.sample(frac=1)
neg_sample = mod_data[mod_data['class_test']==-1][:len(mod_data[mod_data['class_test']==1])]
sample_test = mod_data[mod_data['class_test']==-1][len(mod_data[mod_data['class_test']==1]):]
pos_sample = mod_data[mod_data['class_test']==1]
print(neg_sample.shape, pos_sample.shape)
sample_train = pd.concat([neg_sample, pos_sample]).sample(frac=1)

(77, 27) (77, 27)


In [137]:
classifier = DecisionTreeClassifier()

classifier.fit(sample_train.iloc[:,:-2].values, 
          sample_train.iloc[:,-2].values)
y_predict = classifier.predict(sample_test.iloc[:,:-2].values)
evaluate_results(sample_test.iloc[:,-2].values, y_predict)

Classification results:
f1: 99.75%
roc: 99.75%
recall: 99.51%
precision: 100.00%
